### PR TITLE
Big Cleanup of Java

### LINKED ISSUE #1484: Remove onchain-secrets from java
We still have onchain-secrets in java, this should be removed. When doing that, we also need to keep the `Document` and perhaps the `DecryptKey` classes.

### PR DESCRIPTION
**Pull request by ineiti**
This is a big cleanup of the Java code with regards to removing the OCS library and making more unified classes:
- deleted OCS directory
- ported test using OCS to new ByzCoin / Calypso
- made instances being able to instantiate depending instances
- documented more methods

Closes #1484

### CODE REVIEW

File: byzcoin/proto.go [hunk 1]
```diff
@@ -113,6 +113,26 @@ type GetProofResponse struct {
 	Proof Proof
 }
 
+// CheckAuthorization is used to get the list of actions authorized by a list of
+// identities.
```
Comment (kc1212) [inline byzcoin/proto.go]: I think the comment can be improved. Maybe say "return the list of actions that could be executed if the signatures of the given identities are present and valid".

File: darc/darc.go [hunk 2]
```diff
@@ -822,6 +839,7 @@ func (idkc IdentityX509EC) Verify(msg, s []byte) error {
 	if err != nil {
 		return err
 	}
+	log.Print("verifying 509 signature")
```
Comment (kc1212) [inline darc/darc.go]: remove? or change to a `log.Lvl*`?

File: external/java/src/main/java/ch/epfl/dedis/byzcoin/ByzCoinRPC.java [hunk 3]
```diff
@@ -83,18 +91,22 @@ public ByzCoinRPC(Roster r, Darc d, Duration blockInterval) throws CothorityExce
      * @param skipchainId the ID of the genesis skipblock, aka skipchain ID
      * @throws CothorityException
      */
-    public ByzCoinRPC(Roster roster, SkipblockId skipchainId) throws CothorityException, InvalidProtocolBufferException {
+    public ByzCoinRPC(Roster roster, SkipblockId skipchainId) throws CothorityException{
```
Comment (kc1212) [inline external/java/src/main/java/ch/epfl/dedis/byzcoin/ByzCoinRPC.java]: In the instances, we have `from*` when connecting to an existing instances, the constructor is only used for creating new instances. Do we want something similar for `ByzCoinRPC` too? This one is connecting to an existing skipchain ID, maybe use `fromExisting(Roster roster, SkipblockId skipchainId)`?

File: external/java/src/main/java/ch/epfl/dedis/byzcoin/DataBody.java [hunk 4]
```diff
@@ -1,6 +1,7 @@
 package ch.epfl.dedis.lib.byzcoin;
 
-import ch.epfl.dedis.proto.ByzCoinProto;
+import ch.epfl.dedis.lib.byzcoin.transaction.TxResult;
```
Comment (kc1212) [inline external/java/src/main/java/ch/epfl/dedis/byzcoin/DataBody.java]: this file doesn't have any changes other than an extra import, maybe we don't need it?

File: external/java/src/main/java/ch/epfl/dedis/byzcoin/README.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+Whenever a ClientTransaction is sent to ByzCoin, it contains one or more instructions. Every instruction is sent
+to an instance and can be one of the following:
+
+- spawn - creating a new instance
+- invoke - calling a method of the contract responsible of the instance
+- delete - removing the instance
+
+The first instance in a system is the GenesisDarc that is created and stored in the first block of ByzCoin. This
+genesis darc can be evolved (invoke:evolve) to have more rules that allow to create other instance-types. To do this,
```
Comment (kc1212) [inline external/java/src/main/java/ch/epfl/dedis/byzcoin/README.md]: "more rules that allow to create other instance-types" -> "more rules to create other instance-types"

File: external/java/src/main/java/ch/epfl/dedis/byzcoin/README.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+two methods exist in a `DarcInstance` object:
+
+- `spawnDarc` to create another darc that will be able to receive instructions
+- `spawnContract` to create a general instance of type `ContractId`, where the original darc needs a rule "spawn:ContractId"
```
Comment (kc1212) [inline external/java/src/main/java/ch/epfl/dedis/byzcoin/README.md]: maybe we should change "spawnContract" to something else, it sounds like we can dynamically create contracts. maybe "spawnInstance"

File: external/java/src/main/java/ch/epfl/dedis/byzcoin/contracts/DarcInstance.java [hunk 7]
```diff
@@ -205,10 +221,31 @@ public Proof spawnContractAndWait(String contractID, Signer s, List<Argument> ar
                 throw new CothorityCommunicationException("this is not a correct darc-spawn");
             }
         }
-        logger.info("waiting on iid {}", iid);
         return bc.getProof(iid);
     }
 
+    /**
+     * Spawns a new darc from the given darcInstance. The current darc needs to have the "spawn:darc"
+     * rule with an expression that is resolvable with the signer s.
```
Comment (kc1212) [inline external/java/src/main/java/ch/epfl/dedis/byzcoin/contracts/DarcInstance.java]: "resolvable" might not be the right word, perhaps "...with an expression that can be evaluated to true using signer s"?

File: external/java/src/main/java/ch/epfl/dedis/calypso/CALYPSO.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+Some notes on how to use Calypso in ByzCoin. Calypso allows you to do the following:
+
+Store encrypted data together with a decryption key on ByzCoin, in a way that the decryption key is only accessible
+by trusted parties and only after they authenticated to ByzCoin and proved that they have access.
+
+Given a document with a text `secret` that needs to be put on ByzCoin, the following steps are needed:
+
+1. create a symmetric key and encrypt the `secret` using this symmetric key
+2. encrypt the symmetric using the Long Term Secret public key
+3. send the encrypted `secret` and the encrypted symmetric key to ByzCoin, together with a list of access control rules
```
Comment (kc1212) [inline external/java/src/main/java/ch/epfl/dedis/calypso/CALYPSO.md]: we send the document and the encrypted key to ByzCoin? I thought it's only the encrypted key

File: external/java/src/main/java/ch/epfl/dedis/calypso/CALYPSO.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+by trusted parties and only after they authenticated to ByzCoin and proved that they have access.
+
+Given a document with a text `secret` that needs to be put on ByzCoin, the following steps are needed:
+
+1. create a symmetric key and encrypt the `secret` using this symmetric key
+2. encrypt the symmetric using the Long Term Secret public key
+3. send the encrypted `secret` and the encrypted symmetric key to ByzCoin, together with a list of access control rules 
+
+Now the document is stored on ByzCoin and is ready to be read by anybody being in the list of access control rules:
+
```
Comment (kc1212) [inline external/java/src/main/java/ch/epfl/dedis/calypso/CALYPSO.md]: we need a step for getting the write-proof in the beginning

File: external/java/src/main/java/ch/epfl/dedis/calypso/CalypsoRPC.java [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+     */
+    private CalypsoRPC(Roster roster, SkipblockId byzcoinId, LTSId ltsId) throws CothorityException {
+        super(roster, byzcoinId);
+        Point X = getSharedPublicKey(ltsId);
+        lts = new LTS(ltsId, X);
+    }
+
+    /**
+     * returns the shared symmetricKey of the DKG that must be used to encrypt the
+     * symmetric encryption symmetricKey. This will be the same as OnchainSecretsRPC.X
```
Comment (kc1212) [inline external/java/src/main/java/ch/epfl/dedis/calypso/CalypsoRPC.java]: no more `OnchainSecretsRPC`

File: external/java/src/main/java/ch/epfl/dedis/calypso/CalypsoRPC.java [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    /**
+     * Connects to an existing byzcoin and an existing Long Term Secret.
+     *
+     * @param roster    the nodes handling the byzcoin ledger
+     * @param byzcoinId the id of the byzcoin ledger to connect to
+     * @param ltsId     the id of the Long Term Secret to use
+     * @return CalypsoRPC if everything was found
+     * @throws CothorityException
+     */
+    public static CalypsoRPC fromCalypso(Roster roster, SkipblockId byzcoinId, LTSId ltsId) throws CothorityException {
```
Comment (kc1212) [inline external/java/src/main/java/ch/epfl/dedis/calypso/CalypsoRPC.java]: if we change how byzcoin constructors work (use `fromExisting` to connect to existing services), we should also change this.

File: external/java/src/main/java/ch/epfl/dedis/calypso/LTSId.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
+package ch.epfl.dedis.lib.calypso;
+
+import ch.epfl.dedis.lib.Sha256id;
+import ch.epfl.dedis.lib.exception.CothorityCryptoException;
+import com.google.protobuf.ByteString;
+
+/**
+ * This class represents a SkipblockId, which is a sha256-hash of
+ * the static fields of the skipblock.
```
Comment (kc1212) [inline external/java/src/main/java/ch/epfl/dedis/calypso/LTSId.java]: update comment

File: external/java/src/main/java/ch/epfl/dedis/calypso/ReadData.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+import ch.epfl.dedis.lib.crypto.Point;
+import ch.epfl.dedis.lib.exception.CothorityNotFoundException;
+import ch.epfl.dedis.lib.proto.Calypso;
+import com.google.protobuf.InvalidProtocolBufferException;
+
+/**
+ * A ReadData is the data that is sent to the calypsoRead contract. It is used to log a read request
+ * and must be linked to a corresponding write request.
+ */
+public class ReadData {
```
Comment (kc1212) [inline external/java/src/main/java/ch/epfl/dedis/calypso/ReadData.java]: why not call this `Read`? to be consistent with our other protobuf wrappers

File: external/java/src/main/java/ch/epfl/dedis/calypso/WriteData.java [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+/**
+ * A WriteData is the data that is sent to the calypsoWrite contract store a write request with the encrypted document.
+ * Stored on BzyCoin, it will have the following fields:
+ * <p>
+ * - EncData - the encrypted data, should be smaller than 8MB
+ * - U, Ubar, E, F, Cs - the symmetric key used to encrypt the data, itself encrypted to the Long Term Secret key
+ * - ExtraData - plain text data that is stored as-is on the ledger
+ * - LTSID - the Long Term Secret ID used to encrypt the data
+ */
+public class WriteData {
```
Comment (kc1212) [inline external/java/src/main/java/ch/epfl/dedis/calypso/WriteData.java]: similar to `ReadData`, why not call it `Write`?

File: external/java/src/main/java/ch/epfl/dedis/calypso/CalypsoRPC.java (hunk 11 above)
Comment (ineiti) [reply external/java/src/main/java/ch/epfl/dedis/calypso/CalypsoRPC.java]: I suppose you're referencing the `ByzCoinRPC` constructor.

File: darc/darc.go (hunk 2 above)
Comment (ineiti) [reply darc/darc.go]: Also removed all other `log.Print` in the code.

**Pull request merged**